code/parachain/frame/composable-maths/src/dex/constant_product.rs

```diff
@@ -86,6 +86,59 @@ where
 	Ok(ao)
 }
 
+/// Compute the amount of the output token given the amount of the input token.
+///
+/// Returns a tuple containing `(a_out, fee)`.
+/// To get `a_out` without accounting for the fee, set `f = 0`.
+///
+/// **NOTE:** Weights must be normalized.
+///
+/// From https://balancer.fi/whitepaper.pdf, equation (15)
+///
+/// # Paramaters
+/// * `w_i` - Weight of the input token
+/// * `w_o` - Weight of the output token
+/// * `b_i` - Balance of the input token
+/// * `b_o` - Balance of the output token
+/// * `a_sent` - Amount of the input token sent by the user
+/// * `f` - Total swap fee
+pub fn compute_out_given_in_new<T: PerThing>(
```

Comment:
Need unit tests

Reply:
bf3e802daebf5e1a7f5a8b6f200232e407955e78
Also going to add a test for divide by zero check

Reply:
922cdbdf021ee84d8adeb3c02a61c69ed0627de8